### ISSUE TITLE
Is this project production ready?

### ISSUE THREAD

**Issue by hronro** [shadowsocks/shadowsocks-rust#140]
Thanks for your great jobs to create this amazing project and I really want to use this for my daily use but I'm worrying about the stability. And I would like to know if there is any benchmarks compare to other shadowsocks implementations like `shadowsocks-libev` and `shadowsocks-go`?

Thanks again for your great jobs.

**Comment by cccooogit** [comment]
According to this article :[Making the Tokio scheduler 10x faster](https://tokio.rs/blog/2019-10-scheduler/) ,is there a significant improvement in shadowsocks-rust performance?

**Comment by zonyitoo** [comment]
I am still working on migrating it to tokio v0.2.0-alpha.

**Comment by cg31** [comment]
Adding a tunnel mode will be a help to find the bottleneck.

**Comment by zonyitoo** [comment]
#125

**Comment by Kein** [comment]
Is the local ssslocal client not compatible with server on shadowsock-libev? That would be odd, they use the same shadowsocks proxy protocol.

**Comment by zonyitoo** [comment]
It shouldn't. I have tested it on my own laptop. Could you provide more logs or details?

**Comment by lucidBrot** [comment]
> performance was not my priority goal.
I'm just curious, sorry for hijacking the thread momentarily, but what was your priority goal?

I see so many different implementations for `shadowsocks` that I have to wonder why so many exist.

**Comment by Kein** [comment]
Why not? The more the merrier is the choice.

**Comment by zonyitoo** [comment]
> I'm just curious, sorry for hijacking the thread momentarily, but what was your priority goal?

TL;DR:

1. To make a complete shadowsocks implementation with as less flaws as possible, which contains some useful features for better accessibility for personal usage.
2. To test and verify the networking abilities of Rust's eco-systems.
3. Should be stable for everyday usage.

I started this project when there was just one shadowsocks implementation, the original Python version. When you are trying to deploy a Python program, you have to firstly install a Python, and then make a virtual environment, install dependencies, ... It was so complicated. What I was trying to do was to make server (local and remote) as standalone binaries. There was many choices for me back then, Go, C++, Crystal, D, Rust, ....

I was studying Rust at that time, as a student, so I used it as my first Rust network programming project. After that, I tried to make a coroutine library (runtime) for Rust, and used it in shadowsocks-rust. At that time, this project was a labortory for me to test or verify ideas about networking, coroutine scheduling, and other theories.

In a word, shadowsocks-rust was originally a lab project for testing and verifying Rust's coroutine and networking abilities. After Rust's `async` feature was stablized, and `tokio` project was announced, I have migrated to these new features in Rust and try to test and verify them with this project.

I gave this project to shadowsocks' community because I saw there was so many different implementations out there and all of them have their own characteristics, for example, the libev version is light but lack of multithreading support and multi-servers support, the Go version is "simple" but lack of "design". So my Rust implementation seems unique, which is worth to be put together with other implementations for discussion.

I made this project when I was a student and I was making it just for myself. shadowsocks was originally designed for personal usage, as far as I know. "Performance" shouldn't be the "main" goal for personal users. For them, myself for example, may be more concerned about completeness of features, implementation flaws, and accessibility.

I am totally open for you guys to make this project to be actually "production ready", which involves lots of works to do, such as security reviews, performance improvements, service managements, quality assurance, ... which are all fun subjects to be discussed. But I don't have time to push these forward, also I don't have a "production" environment to verify if they are actually work. It would be irresponsible to say that this project is "production ready".
